src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/cluster/BaseClusterRca.java

```diff
@@ -0,0 +1,227 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+package com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.store.rca.cluster;
+
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.framework.api.Rca;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.framework.api.Resources;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.framework.api.Resources.State;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.framework.api.contexts.ResourceContext;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.framework.api.flow_units.ResourceFlowUnit;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.framework.api.summaries.HotClusterSummary;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.framework.api.summaries.HotNodeSummary;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.scheduler.FlowUnitOperationArgWrapper;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.reader.ClusterDetailsEventProcessor;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.reader.ClusterDetailsEventProcessor.NodeDetails;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.HashBasedTable;
+import com.google.common.collect.Table;
+import java.time.Clock;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * This is a generic cluster level RCA which subscripts a single upstream node level RCA.
+ * This cluster RCA maintains a Table to keep track of flowunits sending from different nodes across
+ * the cluster. This table is a two dimensional table indexed by (NodeKey, Rca Name). This RCA will
+ * mark the cluster as unhealthy if the flowunits from any data nodes are unhealthy.
+ * <p></p>
+ * A few protected variables that can be overridden by derived class:
+ * numOfFlowUnitsInMap : number of consecutive flowunits stored in hashtable. Default is 1
+ * collectFromMasterNode : whether this RCA collect flowunit from master nodes.
+ * expirationTimeWindow : time window to determine whether flowunit in hashmap becomes stale
+ */
+public class BaseClusterRca extends Rca<ResourceFlowUnit<HotClusterSummary>> {
```

Comment:
If we are building a base class for all RCA types, can we keep the `HotClusterSummary` part generic so that TemperatureProfile RCAs can make use of it as well ?

Reply:
As per out discussion, I will create separate base class for cluster summary and node summary so that we can make the  output flowunit generic. Let me address it in a separate PR. will create a issue for that